src/auth/auther/Auther.ts

```diff
@@ -0,0 +1,88 @@
+import { AuthResult } from './AuthResult'
+import { notFound, redirect } from 'next/navigation'
+import type { SessionMaybeUser, SessionUser } from '@/auth/Session'
+
+type AutherRedirectConfig = {
+    shouldRedirect: boolean,
+    redirectUrl: string | null,
+    returnUrl: string | null,
+}
+
+export type UserRequieredOutOpt = 'USER_NOT_REQUIERED_FOR_AUTHORIZED' | 'USER_REQUIERED_FOR_AUTHORIZED'
+
+export type CheckReturn<
+    UserRequieredOut extends UserRequieredOutOpt
+> = UserRequieredOut extends 'USER_REQUIERED_FOR_AUTHORIZED' ? ({
+    success: true,
+    session: SessionUser
+} | {
+    success: false,
+    session: SessionMaybeUser
+}) : ({
+    success: boolean,
+    session: SessionMaybeUser
+})
+
+export abstract class Auther<
+    const UserRequieredOut extends UserRequieredOutOpt,
+    DynamicFields,
+> {
+    private redirectConfig: AutherRedirectConfig
+
+    public constructor() {
+        this.redirectConfig = {
+            shouldRedirect: false,
+            redirectUrl: null,
+            returnUrl: null
+        }
+        return this
+    }
+
+    protected abstract check(
+        session: SessionMaybeUser,
+        dynamicFields: DynamicFields
+    ): CheckReturn<UserRequieredOut>
+
+    public auth({
+        session: sessionIn,
+        dynamicFields
+    }: {
+        session: SessionMaybeUser,
+        dynamicFields: DynamicFields
```

Comment:
Why isn't this an optional field? The syntax `dynamicField: undefined` is a bit ugly

Reply:
It is very hard to get ts to agree to that. I will look info it when I change it to be just functional